### PR TITLE
chore: replace wsconncache with a single tailnet

### PR DESCRIPTION
**Pull request by coadler**
This PR mostly removes `wsconncache` from coderd, instead replacing it with a single Tailnet per coderd. The biggest benefit to this is not needing a cache of Tailnets sitting around for each agent. Additionally, by reusing the same `http.Transport` for proxying to workspaces, we're able to reuse TCP conns between requests instead of them being open and closed after each request. This is the same strategy we use in wgtunnel.

I made some changes to the coordinator interface to support coderd hooking directly in, so I'd definitely like some feedback on that. I haven't yet implemented it for pgcoord.

Left out of this PR is support for moons. This was getting a bit too big, so I'll instead do that in a followup. Moons will still use wsconncache for the time being.

### CODE REVIEW

File: agent/agenttest/client.go
Comment (coadler) [inline agent/agenttest/client.go:1]: This was essentially copy/pasted out of `agent_test.go` so it can be reused.

File: coderd/tailnet.go
Comment (kylecarbs) [inline coderd/tailnet.go:111]: Should we toss this in the `tailnet` package for now? Or can we not because of the `wsconncache` dep?

File: coderd/tailnet.go
Comment (kylecarbs) [inline coderd/tailnet.go:42]: Although maybe a bit weird, I'd argue we should put this in `workspaceagents.go` since `wsconncache` will be going away.

File: coderd/tailnet.go
Comment (kylecarbs) [inline coderd/tailnet.go:28]: Because this is global lets do `tailnetTransport` or something

File: coderd/workspaceagents.go
Comment (kylecarbs) [inline coderd/workspaceagents.go:732]: When will this be removed? Let's link an issue or something.

File: codersdk/workspaceagentconn.go
Comment (kylecarbs) [inline codersdk/workspaceagentconn.go:138]: Should this all be in `agentsdk` instead? Seems like a better spot

**Review by kylecarbs — APPROVED** [review]
I didn't manually test this, but the code looks good. It's sick that it's backward-compatible.

We should have a timeline on the removal of wsconncache, because that seems like some big dead weight.

File: coderd/tailnet.go
Comment (coadler) [reply coderd/tailnet.go:111]: Yeah, putting this in `tailnet` causes cyclical imports

Comment (kylecarbs) [reply coderd/tailnet.go:111]: Reasonable

File: codersdk/workspaceagentconn.go
Comment (coadler) [reply codersdk/workspaceagentconn.go:138]: yeah i think that makes more sense

File: codersdk/workspaceagents.go
Comment (mafredri) [inline codersdk/workspaceagents.go:296]: This message could potentially lie since we're checking len == 1.

File: coderd/tailnet_test.go
Comment (mafredri) [inline coderd/tailnet_test.go:186]: There's no error to use here, comes from parent scope.

**Review by spikecurtis — CHANGES_REQUESTED** [review]

File: enterprise/tailnet/coordinator.go
Comment (coadler) [reply enterprise/tailnet/coordinator.go:369]: I've modified the interface so these callbacks are no longer necessary. Impressive find :1st_place_medal:

**Comment by coadler** [comment]
Ignore test failures, will fix those in the morning.

File: tailnet/multiagent.go
Comment (spikecurtis) [inline tailnet/multiagent.go:60]: it looks like this only gets called when an agent publishes a new node, but the MultiAgentConn wants to get the latest node bindings for agents as soon as they subscribe---not wait until the agent publishes an update, which could be never, if its endpoints and DERP home haven't changed.

File: tailnet/multiagent.go
Comment (spikecurtis) [inline tailnet/multiagent.go:105]: I don't think you want this `default:` case here, which allows the `NextUpdate()` to return if there isn't an update ready.

The zero value of `nodes` is `nil`, and in `watchAgentUpdates()` if it gets `nil` it exits.  If you return an empty slice then `watchAgentUpdates()` will go into a tight loop doing nothing but eating up CPU.

What you want is for `NextUpdate()` to _block_ until an update is available (or context expires), so drop the `default:` case.

---
As an aside, 

```
loop:
    for {
        select {
        ...
        default:
            break loop
        }
    }
    return nodes
```

can be simplified to

```
    for {
        select {
        ...
        default:
            return nodes
        }
    }
```

File: enterprise/tailnet/pgcoord.go
Comment (spikecurtis) [inline enterprise/tailnet/pgcoord.go:128]: this effectively torpedos the `pgCoord`, so we can't merge this PR until fixed.

**Review by spikecurtis — CHANGES_REQUESTED** [review]

**Comment by spikecurtis** [comment]
FWIW, I think the right architecture for both the in-memory and distributed coordinator is to consider an individual client connection as a special case of the more general "multiagent" idea.  That is, a normal end user client is a multiagent where the number of agents is exactly one.

It's much easier to maintain and understand code that computes the general case, and then at the edge, we have adaptations the wire up clients with exactly one agent, rather than trying to build two different sets of logic and keep them consistent.

So, in the core logic of the coordinator, clients are all multi-agent, and we compute and queue up updates on them based on the set of agents they've asked to connect to.

Then, at the very edge we have special case code that either serializes the nodes over the websocket, or sends them out over channels/callbacks depending on whether we have an end user, remote client, or a coderd, in-process client.

File: coderd/tailnet.go
Comment (deansheather) [inline coderd/tailnet.go:152]: This is so nice, great work!

File: codersdk/workspaceagentconn.go
Comment (deansheather) [reply codersdk/workspaceagentconn.go:139]: This shouldn't go in agentsdk, agentsdk is for agent making requests to coderd. `WorkspaceAgentConn` is clients making connections to agents.

File: codersdk/workspaceagents.go
Comment (deansheather) [inline codersdk/workspaceagents.go:298]: Wait really? We can only get one node at a time? Why is it an array to begin with then?

File: tailnet/coordinator.go
Comment (deansheather) [reply tailnet/coordinator.go:471]: Or NoLock which makes more sense IMO and is used in the dbfake package

File: enterprise/tailnet/pgcoord.go
Comment (coadler) [reply enterprise/tailnet/pgcoord.go:128]: Since I'm still mostly flushing out the API I didn't think it made sense to impl on `pgCoord` yet. Definitely a blocker for merge though.

File: tailnet/multiagent.go
Comment (coadler) [reply tailnet/multiagent.go:105]: Ah, I see now. What I was trying to do was pop all of the currently buffered nodes on the channel, and exit once drained. I guess the only way to do this then is by debouncing then.

File: codersdk/workspaceagents.go
Comment (coadler) [reply codersdk/workspaceagents.go:298]: Yeah, I'm not sure why though. This is actually vestigal though so it can be removed

File: tailnet/coordinator.go
Comment (coadler) [inline tailnet/coordinator.go:185]: Havent added this to the `haCoordinator` yet.

File: tailnet/coordinator.go
Comment (spikecurtis) [inline tailnet/coordinator.go:549]: We still have this architecture where, even in the `core`, clients with a single agent and multiple agents are being handled independently.  From the perspective of the `core`, they should be identical.

In the code now, we've got this inconsistent handling, where for single-agent clients, the `core` is keeping track of the set (of size 1) of agents that the client is connecting to, and the core is governed by a mutex that makes it fairly easy to ensure that the right updates get sent out as things change.

However, for multi-agent clients, the core doesn't know all the agents the client is connecting to (technically it could compute it by inverting the map indexes on `agentToConnectionSockets`), and it instead depends on the `MultiAgent` to keep track and tell it.  But, this has resulted in a race condition bug that goes like this:

1. ServerTailnet calls `UpdateSelf()`, and we collect all the agents we should send an update to, and unlock the `MultiAgent` mutex.
2. In another goroutine, we call `SubscribeAgent()` with a new Agent.  This acquires the `MultiAgent` mutex, and then calls `core.MultiAgentSubscribe()`.  This locks the core, and sends the agent the stale node data (or, no node if step 1 above is the first time UpdateSelf() is being called).
3. The core unlocks, and the goroutine from step 1 calls `core.multiAgentUpdate()`.  However the agent from step 2 is not in the list, and it never gets the update.  Sadface.

I really think the core should be written to _only_ deal with clients that can have multiple agents, and the core should be responsible for all the state about who is connecting to who.  End-user clients that only have one agent should be modeled in the core as an initialization followed immediately by a subscribe.

You'll probably want to have a map that's the inverse of `agentToConnectionSockets`, where the first index is clientID and the second index is `agentID` --- this will make client node updates fast, since now we won't have the agentID on calls to `clientNodeUpdate()`

**Review by spikecurtis — CHANGES_REQUESTED** [review]

File: enterprise/tailnet/pgcoord.go
Comment (spikecurtis) [reply enterprise/tailnet/pgcoord.go:128]: FWIW, I think what we need to do is to drop the `agent_id` column from `tailnet_clients` table, and add a new table that tracks the agents that each client wants to connect to

```sql
CREATE TABLE tailnet_subscriptions (
    client_id uuid NOT NULL,
    coordinator_id uuid NOT NULL,
    agent_id uuid NOT NULL,
    PRIMARY KEY (client_id, coordinator_id, agent_id),
    FOREIGN KEY (client_id) REFERENCES tailnet_clients(id) ON DELETE CASCADE,
    FOREIGN KEY (coordinator_id) REFERENCES tailnet_coordinators(id) ON DELETE CASCADE
);
```

File: tailnet/multiagent.go
Comment (coadler) [reply tailnet/multiagent.go:108]: I guess I was mostly trying to avoid footguns when closing, and also allow for separate stat tracking. It seemed confusing/surprising to me that closing a single `Enqueuable` could have effects beyond that instance, possibly closing other `Enqueuable`s, especially when we're trying to treat them as equals to regular connections.

Comment (spikecurtis) [reply tailnet/multiagent.go:108]: But, what does it mean to "close" one of these?

Closing a queue should mean that you can't enqueue anything else on it and that the other side should get notified that it's closed.  That needs to be done at the level of the `MultiAgent` as a whole, not just in the context of receiving updates for a single agent.

File: coderd/tailnet.go
Comment (coadler) [inline coderd/tailnet.go:133]: Will do this before merge

File: tailnet/coordinator.go
Comment (spikecurtis) [inline tailnet/coordinator.go:264]: I think it's better to continue to use a `TrackedConn` for remote clients - the power of the `Enqueuable` abstraction is that the `core` can treat it equivalently to `MultiAgent`.

Then we won't need `SendUpdatesToConn()` as a separate function that essentially repeats most of `TrackedConn.SendUpdates()`.

File: tailnet/trackedconn.go
Comment (spikecurtis) [inline tailnet/trackedconn.go:20]: Good call on moving this stuff to its own file.  Much more readable.

File: tailnet/coordinator.go
Comment (spikecurtis) [inline tailnet/coordinator.go:200]: I don't understand the distinction between `CoordinatorClose()` and `Close()` --- how are they different and why do we need both?

File: tailnet/coordinator.go
Comment (spikecurtis) [inline tailnet/coordinator.go:180]: this could be a `map[uuid.UUID]map[uuid.UUID]Enqueueable` and then when you get a client update, you don't have to do a separate lookup in the `agentSockets` to get the socket.

File: tailnet/coordinator.go
Comment (spikecurtis) [inline tailnet/coordinator.go:193]: "Enqueuable" is hard to say and type.  Could we just call this a "Queue"?

File: tailnet/multiagent.go
Comment (spikecurtis) [inline tailnet/multiagent.go:61]: closing the `MultiAgent` feels incomplete to me since it seems that we would still send updates, subscribes, and unsubscribes into the `core`, even after it's been closed.

Furthermore, `NextUpdate()` runs on is own, separate Context, rather than respecting the closed state of the `MultiAgent`.  This seems like a recipe for zombie goroutines.

**Review by spikecurtis — CHANGES_REQUESTED** [review]

File: tailnet/multiagent.go
Comment (coadler) [reply tailnet/multiagent.go:61]: `NextUpdate()` does respect the closed state of `MultiAgent` since `m.updates` gets closed when the `MultiAgent` is closed.

File: tailnet/coordinator.go
Comment (coadler) [reply tailnet/coordinator.go:199]: There's sort-of a chicken and egg problem. When you want to close the individual queue, you expect it to be removed from the coordinator. When the coordinator is closing, you want the coordinator to close all of the queues, but you don't want the queues to reach back into the coordinator to close themselves and reenter the mutex.

**Comment by coadler** [comment]
I've made this opt-in via a feature flag, so it won't be enabled by default. The implementations for pgcoord and moons aren't done, but I plan on addressing that in follow-up PRs due to this one getting huge. After both are resolved, I'll remove the feature flag.

File: coderd/tailnet.go
Comment (spikecurtis) [inline coderd/tailnet.go:126]: This measures the last time we _started_ a connection, not the last time the connection was used.  If we proxy a long-lived connection like `ReconnectingPTY` or a devURL websocket, it could easily be in use for greater than 30 minutes.

We might need some ref-counting to keep track of the connections to each agent, so that we expire them when they are no longer used.

File: tailnet/multiagent.go
Comment (spikecurtis) [inline tailnet/multiagent.go:81]: This can cause a deadlock if something attempts to `Close()` the multi-agent.

> If a goroutine holds a RWMutex for reading and another goroutine might call Lock, no goroutine should expect to be able to acquire a read lock until the initial read lock is released. In particular, this prohibits recursive read locking. This is to ensure that the lock eventually becomes available; a blocked Lock call excludes new readers from acquiring the lock.
>
> https://pkg.go.dev/sync#RWMutex

`m.OnSubscribe()` can result in a call back to `Enqueue()` if the node for the agent is known, resulting in reentrant call to `RLock()`.

**Review by spikecurtis — APPROVED** [review]
I understand this PR is getting big and unwieldy, so I'm fine with it going in behind an experiment flag, but I would like to see a GitHub epic and GitHub issues tracking the things that need a resolution before this can go GA.

**Pull request merged**
